module/zfs/zfs_ioctl.c

```diff
@@ -5232,25 +5499,992 @@ zfs_ioc_send_space(const char *snapname, nvlist_t *innvl, nvlist_t *outnvl)
+	    CRED(), &resid);
+
+	areleasef(dls->dls_fd, dls->dls_fip);
+	kmem_free(dls->dls_fsname, sizeof (dls_t));
+	kmem_free(dls, sizeof (dls_t));
+}
+
+static int
+zfs_stable_ioc_zfs_list(const char *fsname, nvlist_t *innvl,
+    nvlist_t *outnvl, nvlist_t *opts, uint64_t version)
+{
+	dsl_pool_t *dp = NULL;
+	int fd;
+	file_t *fp;
+	nvlist_t *nv;
+	dls_t *dls;
+	dls_flag_t dls_flags = 0;
+	uint64_t depth = 0-1;
+	uint64_t dd_obj = 0;
+	int error;
+
+	error = nvlist_lookup_int32(innvl, "fd", &fd);
+	if (error != 0)
+		return (SET_ERROR(EINVAL));
+
+	if (nvlist_exists(opts, "recurse")) {
+		dls_flags |= DLS_RECURSE;
+		(void) nvlist_lookup_uint64(opts, "recurse", &depth);
+	}
+
+	if (nvlist_lookup_nvlist(opts, "type", &nv) == 0) {
+		if (nvlist_exists(nv, "all"))
+			dls_flags |= DLS_TRAVERSE_ALL;
+		else {
+			if (nvlist_exists(nv, "bookmark"))
+				dls_flags |= DLS_TRAVERSE_BOOKMARK;
+			if (nvlist_exists(nv, "filesystem"))
+				dls_flags |= DLS_TRAVERSE_FILESYSTEM;
+			if (nvlist_exists(nv, "snap"))
+				dls_flags |= DLS_TRAVERSE_SNAPSHOT;
+			if (nvlist_exists(nv, "snapshot"))
+				dls_flags |= DLS_TRAVERSE_SNAPSHOT;
+			if (nvlist_exists(nv, "volume"))
+				dls_flags |= DLS_TRAVERSE_VOLUME;
+		}
+	}
+
+	/* Pass an object number when given a DSL directory name */
+	if (fsname != NULL && fsname[0] != '\0') {
+		spa_t *spa;
+		dsl_dataset_t *ds = NULL;
+
+		mutex_enter(&spa_namespace_lock);
+
+		spa = spa_lookup(fsname);
+		if (spa == NULL) {
+			mutex_exit(&spa_namespace_lock);
+			return (ENOENT);
+		}
+
+		dp = spa_get_dsl(spa);
+
+		dsl_pool_config_enter(dp, FTAG);
+		mutex_exit(&spa_namespace_lock);
+
+		error = dsl_dataset_hold(dp, fsname, FTAG, &ds);
+
+		/* Adopt sane defaults based on the DSL directory */
+		if (!nvlist_exists(nv, "type")) {
+			if ((strchr(fsname, '#') != NULL)) {
+				dls_flags |= DLS_TRAVERSE_BOOKMARK;
+				dls_flags &= ~DLS_RECURSE;
+			} else if ((strchr(fsname, '@') != NULL)) {
+				dls_flags |= DLS_TRAVERSE_SNAPSHOT;
+				dls_flags &= ~DLS_RECURSE;
+			} else {
+				objset_t *osp;
+				dmu_objset_from_ds(ds, &osp);
+
+				switch (dmu_objset_type(osp)) {
+				case DMU_OST_ZVOL:
+					dls_flags |= DLS_TRAVERSE_VOLUME;
+					break;
+				case DMU_OST_ZFS:
+					dls_flags |= DLS_TRAVERSE_FILESYSTEM;
+					break;
+				default:
+					VERIFY(0);
+				}
+			}
+		}
+
+		dsl_pool_config_exit(dp, FTAG);
+		if (error)
+			return (error);
+		dd_obj = ds->ds_dir->dd_object;
+		dsl_dataset_rele(ds, FTAG);
+	}
+
+	fp = getf(fd);
+	if (fp == NULL) {
+		if (dp)
+			dsl_pool_config_exit(dp, FTAG);
+		return (SET_ERROR(EBADF));
+	}
+
+	dls = kmem_alloc(sizeof (dls_t), KM_SLEEP);
+	dls->dls_fd = fd;
+	dls->dls_dp = dp;
+	dls->dls_dd_obj = dd_obj;
+	dls->dls_fip = P_FINFO(curproc);
+	dls->dls_vp = fp->f_vnode;
+	dls->dls_depth = depth;
+	dls->dls_flags = dls_flags;
+	dls->dls_fsname = (fsname != NULL && fsname[0] != '\0') ?
+	    strdup(fsname) : NULL;
+
+	if (!taskq_dispatch(system_taskq, dump_list_strategy, dls, TQ_SLEEP)) {
```

Comment:
It is natural for API for `zfs send` to be called in a different thread than the API for `zfs recv`, but it is not natural for the API for `zfs list` to be called in a different thread than the code that reads it. I believe that doing this in the same thread would be prone to deadlocks from incorrect usage and I did not like the idea of subjecting the `spa_namespace` lock to that risk. With `zfs send`, the risk is limited to the snapshot being sent.

That said, it occurs to me that unprivileged users with access to this API could perform denial of service whenever it is possible for the pipe buffer to fill. That is not a problem on Linux where only root can access this API, but it would be a big problem for Illumos and FreeBSD. I will need to think of a way to avoid that. Doing CoW inside the kernel itself on the state of the SPA and DSL would make this work, although I would appreciate other ideas.